anxcloud/common_resource_tagging.go

```diff
@@ -0,0 +1,106 @@
+package anxcloud
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sort"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"go.anx.io/go-anxcloud/pkg/api"
+	corev1 "go.anx.io/go-anxcloud/pkg/apis/core/v1"
+)
+
+func withTagsAttribute(s schemaMap) schemaMap {
+	s["tags"] = &schema.Schema{
+		Type:        schema.TypeList,
+		Optional:    true,
+		Computed:    true,
+		Description: "List of tags attached to the resource.",
+		Elem: &schema.Schema{
+			Type: schema.TypeString,
+		},
+		// suppress diff when only the order has changed
+		DiffSuppressOnRefresh: true,
+		DiffSuppressFunc: func(k, oldValue, newValue string, d *schema.ResourceData) bool {
+			o, n := d.GetChange("tags")
+			oStringArray := castInterfaceArray[string](o.([]interface{}))
+			nStringArray := castInterfaceArray[string](n.([]interface{}))
+			sort.Strings(oStringArray)
+			sort.Strings(nStringArray)
+			return reflect.DeepEqual(oStringArray, nStringArray)
+		},
+	}
+	return s
+}
+
+type schemaContextFunc func(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics
+
+func ensureTagsMiddleware(pre schemaContextFunc) schemaContextFunc {
```

Comment:
you already use generics quite often, why not here? Would remove quite a lot of castings in the resource schemas ^^

Reply:
maybe call `pre` something like `wrapped`, I think that fits a bit better?

Reply:
re generics: hm, I was sure golang could infer that.. with now stating the return value in the type argument, it does not really look better than the cast - narf :/

Reply:
so yeah, not sure ... let it stay this way with the generic function or revert to casting? Do you have any idea what's betterß

Reply:
besides this thread lgtm :)

Reply:
oki, then revert to cast version please :)

Reply:
Or maybe `createTagsMiddleware` and `updateTagsMiddleware` functions which call the common function and do the casting?

Reply:
thought about that, too but dismissed this idea .. since you came up with the same, it's probably not as bad though :D

yeah, go for it :+1: